### PR TITLE
Display parsed user agent information in assessment instance log

### LINKED ISSUE #9746: Parse user agent in fingerprint popover
The fingerprint popover shows the raw value of the user agent. It would be nice to show user friendly information by parsing it (e.g., with https://www.npmjs.com/package/ua-parser-js). The raw value could still be shown, but in a faded style.

### PR DESCRIPTION
**Pull request by jonatanschroeder**
Resolves #9746.

image

Bowser was used to parse the data because it was already in use, and because it has a permissive license (MIT). That said, its current use is not actually needed anymore (and was removed). It was needed in the previous version of the mathjax handler that switched from SVG to CHTML depending on the browser, but that is (1) no longer used, and (2) no longer needs the same res.locals processing since we don't rely on EJS anymore.

ua-parser-js was not used because [the latest version (2.0.0) uses AGPL](https://github.com/faisalman/ua-parser-js?tab=readme-ov-file#license-options), which may be fine but could cause an issue in the future. The previous version (1.0) uses MIT, but I'm afraid it may not get as much TLC in the future given the newer version. That said, switching from bowser to ua-parser-js is relatively straightforward if we're ok with their license. One advantage of ua-parser-js is that it is more actively maintained and (claims to) have support for a larger set of browsers, though for our use case I don't really see that as significant.

### CODE REVIEW

File: apps/prairielearn/src/server.js [hunk 1]
```diff
@@ -175,15 +174,6 @@ export async function initExpress() {
   // API routes don't utilize sessions; don't run the session/flash middleware for them.
   app.use(excludeRoutes(['/pl/api'], sessionRouter));
 
-  app.use(function (req, res, next) {
-    if (req.headers['user-agent']) {
-      res.locals.userAgent = Bowser.parse(req.headers['user-agent']);
-    } else {
-      res.locals.userAgent = null;
```
Comment (jonatanschroeder) [inline apps/prairielearn/src/server.js:182]: AFAICT this is the last time this stopped being used: https://github.com/PrairieLearn/PrairieLearn/pull/6974/files#diff-cae3cb6cc3f271a4dd1c43d90e822ceb3d82a56884ffa57728af052ed0aaa284

**Comment by nwalters512** [comment]
AGPL might not be a problem for us? I think it just requires us to release any modifications to it, of which we wouldn't have any (based on reading this: https://drewdevault.com/2020/07/27/Anti-AGPL-propaganda.html). cc @mwest1066 who I know has thoughts about licensing.

**Comment by jonatanschroeder** [comment]
About the Windows 11 issue: it seems this requires additional headers beyond the user agent (see https://learn.microsoft.com/en-us/microsoft-edge/web-platform/how-to-detect-win11 and https://developer.mozilla.org/en-US/docs/Web/API/User-Agent_Client_Hints_API). This also applies to the distinction between "Android" and "Linux", or to distinguish macOS versions above 10.15. `ua-parser-js` has the means to handle these client hints (see https://docs.uaparser.dev/api/main/idata/with-client-hints.html), but this would require:

* adding a couple of additional headers to our responses to ask the browser to provide extra information in future requests (https://docs.uaparser.dev/api/main/idata/with-client-hints.html#create-a-simple-http-server, https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Accept-CH, https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Critical-CH);
* extend the client fingerprints to retrieve and store headers that start with `sec-ch-ua-`;
* pass these extra headers to the parser code (https://docs.uaparser.dev/api/main/idata/with-client-hints.html#server-side-example).

This would only provide additional info in Chromium-based browsers (Chrome, Edge, Opera), since this feature is not currently supported by Firefox and Safari.

**Comment by mwest1066** [comment]
tldr: we shouldn't use GPL/AGPL libraries in PL.

The blog article that @nwalters512 references doesn't address the key question of GPL/AGPL libraries. It seems more focused on using services such as PostgreSQL. I think everyone agrees that, even if PostgreSQL was GPL or AGPL, software using it would not need to be GPL/AGPL, which seems to be the point that this blog post is making.

The case that @jonatanschroeder raises above is about an AGPL library. My understanding about GPL libraries is as follows. The question is: if I write a program that uses a library licensed under a GPL-variant license, does my program also have to be similarly GPL licensed? The summary is:
* GPLv2: somewhat unclear, may depend on static vs dynamic linking, hasn't been tested in court.
* GPLv3: yes, the program must also be GPLv3
* AGPL: yes, the program must also be AGPL
* LGPL: no, the program has no particular license requirements

For NodeJS code using npm libraries, the general understanding seems to be that it's roughly equivalent to dynamic linking in compiled languages, although this has never been tested in court. Because of the dual-licensed EE/CE nature of PL, this means we can't use GPL/AGPL libraries. We would be fine to use LGPL libraries.

More details:
* The basic test in copyright law is whether a work is a "derivative" of another. For software programs the exact definition of "derivative work" was never specified in copyright law, and to my knowledge there has not been any case law that gives a definitive answer here. One key unknown is whether a program is a derivative work of a linked library, and whether the form of that linking (e.g., static vs dynamic) makes a difference.
* GPLv2 used the term "derivative work" and so it is still unclear (pending future case law) exactly how this applies to libraries.
* To address this issue, GPLv3 and AGPL stopped using the term "derivative work" and instead refer explicitly to "shared libraries and dynamically linked subprograms", making it clear that they apply to any works which link the GPLv3/AGPL library, whether statically or dynamically.
* The FSF uses the separate LGPL license for libraries in cases where the license is not intended to apply to programs using the library.

**Comment by nwalters512** [comment]
@jonatanschroeder are you still interested in merging this?

**Comment by jonatanschroeder** [comment]
> @jonatanschroeder are you still interested in merging this?

I am ok with the current strategy, though before I take care of merge conflicts I'd appreciate any thoughts on if this is the way to go. In particular, is there a concern with doing the corner case processing inside PL itself?

**Comment by nwalters512** [comment]
Discussed in the dev meeting yesterday: I'm good to use client hints here. They're more future-proof and can provide more information than what we'd get in `User-Agent`.

**Comment by jonatanschroeder** [comment]
I tested with multiple browsers in my computer. Edge and Chrome seem to be sending client hints as expected. Firefox is not (which [MDN acknowledges](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Accept-CH#browser_compatibility)). I tested connecting from my phone, but because `Accept-CH` is [only available in secure connections](https://developer.mozilla.org/en-US/docs/Web/HTTP/Reference/Headers/Accept-CH) I was unable to get useful info for testing.

I believe the client hints by themselves may be somewhat limited at this point given the lack of support from Firefox and Safari. That said, I don't think there's any harm with keeping the implementation as is for now, providing information where it is available. Another alternative is to fallback to the MIT-licensed version of `ua-parser-js` when client hints are not available.

**Comment by nwalters512** [comment]
> Another alternative is to fallback to the MIT-licensed version of ua-parser-js when client hints are not available.

I lean somewhat towards doing just this for now, maybe instead of client hints entirely. Client hints are interesting and probably hold value in the long run, but just parsing data we're already storing is simpler and lets us avoid DB schema changes, which always makes it easier to justify quickly merging a PR. The MIT-licensed version of `us-parser-js` was most recently published 4 months ago so it should be up-to-date enough, especially for things like Windows 11, which you mentioned Bowser doesn't support.

Another reason to use `User-Agent`: we've already stored it for every fingerprint ever captured so we don't have to worry about missing data for existing fingerprints.

**Comment by jonatanschroeder** [comment]
> The MIT-licensed version of `us-parser-js` was most recently published 4 months ago so it should be up-to-date enough, especially for things like Windows 11, which you mentioned Bowser doesn't support.

It supports more than Bowser, but it does not support Windows 10 vs 11 because that requires client hints, which are not available on the MIT version.

I did some tests to check how bowser and ua-parser-js (v1.0.40) compare with [user agents listed online for popular devices](https://deviceatlas.com/blog/list-of-user-agent-strings). Bowser provides good coverage for a bunch of cases, but there are some specific cases where ua-parser-js provides better info. Examples:

* For Android, ua-parser-js is able to provide a device vendor for several devices, and distinguish tablet vs mobile.
* It can find the version of Chrome OS.
* It lists Ubuntu instead of Linux.
* It can identify some devices like Playstation 5 and Kindle (not a big deal, I don't expect students to write exams in a PS5, but regardless...).

I'm opening a separate PR to use ua-agent-js to retrieve this data.
